File: workbook/ch07/data/common/kmp.py
def compute_prefix_table(pattern):
    pattern_length = len(pattern)
    prefix_table = [0] * pattern_length
    length = 0  # length of previous longest prefix suffix

    for i in range(1, pattern_length):
        while length > 0 and pattern[i] != pattern[length]:
            length = prefix_table[length - 1]

        if pattern[i] == pattern[length]:
            length += 1
            prefix_table[i] = length

    return prefix_table

def kmp_search(text, pattern):
    text_length = len(text)
    pattern_length = len(pattern)
    prefix_table = compute_prefix_table(pattern)

    i = 0  # index for text
    j = 0  # index for pattern

    while i < text_length:
        if pattern[j] == text[i]:
            i += 1
            j += 1

        if j == pattern_length:
            print(f"Pattern found at index {i - j}")
            j = prefix_table[j - 1]
        elif i < text_length and pattern[j] != text[i]:
            if j != 0:
                j = prefix_table[j - 1]
            else:
                i += 1


text = "ABABDABACDABABCABAB"
pattern = "ABABCABAB"

print(f"Text: {text}")
print(f"Pattern: {pattern}")

kmp_search(text, pattern)

File: workbook/ch07/data/common/kmp.c
#include <stdio.h>
#include <string.h>

// compute the prefix table (also called the "failure function")
void computePrefixTable(char *pattern, int patternLength, int *prefixTable) {
    int length = 0;  // length of the previous longest prefix suffix
    prefixTable[0] = 0;  // first value is always 0

    for (int i = 1; i < patternLength; ) {
        if (pattern[i] == pattern[length]) {
            length++;
            prefixTable[i] = length;
            i++;
        } else {
            if (length != 0) {
                length = prefixTable[length - 1];
            } else {
                prefixTable[i] = 0;
                i++;
            }
        }
    }
}

void KMPSearch(char *text, char *pattern) {
    int textLength = strlen(text);
    int patternLength = strlen(pattern);

    int prefixTable[patternLength];
    computePrefixTable(pattern, patternLength, prefixTable);

    int i = 0;  // index for text
    int j = 0;  // index for pattern

    while (i < textLength) {
        if (pattern[j] == text[i]) {
            i++;
            j++;
        }

        if (j == patternLength) {
            printf("Pattern found at index %d\n", i - j);
            j = prefixTable[j - 1];
        } else if (i < textLength && pattern[j] != text[i]) {
            if (j != 0) {
                j = prefixTable[j - 1];
            } else {
                i++;
            }
        }
    }
}

int main() {
    char text[] = "ABABDABACDABABCABAB";
    char pattern[] = "ABABCABAB";

    printf("Text: %s\n", text);
    printf("Pattern: %s\n", pattern);

    KMPSearch(text, pattern);

    return 0;
}
